packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -445,8 +445,10 @@ class AndroidDevice extends Device {
       printError('Error waiting for a debug connection: $error');
       return new LaunchResult.failed();
     } finally {
-      observatoryDiscovery.cancel();
-      diagnosticDiscovery.cancel();
+      await Future.wait(<Future<Null>>[
+        observatoryDiscovery.cancel(),
+        diagnosticDiscovery.cancel(),
+      ].where((Future<Null> future) => future != null));
```

Comment:
we should have a helper function that does all this for you. This isn't code we should be repeating all over the place. :-)